### PR TITLE
Function Overloading

### PR DESCRIPTION
**Pull request by cjmanca**
Added support for function overloading, by specifying a list of
parameter types.

Example:
```javascript
var TestClass = Class('TestClass', {
    x: 6,
    y: "something else"
});
var MyClass = Class('MyClass', {
    a: 1,
    b: "test",

    'test(number)': function(num) {
        this.a = num;
    },
    'test(string)': function(str) {
        this.b = str;
    },
    'test(TestClass)': function(val) {
        this.a = val.getX();
        this.b = val.getY();
    },
    'test(object)': function(val) {
        if ('a' in val) {
            this.a = val.a;
        }
        if ('b' in val) {
            this.b = val.b;
        }
    },
    'test(string, ...)': function(str, args) {
        return str + " = " + args.join('-');
    }
});
var mc = new MyClass();


mc.test(2);
mc.getA(); // returns 2

mc.test("success");
mc.getB(); // returns "success"

mc.test({ a: 5, b: "passed" });
mc.getA(); // returns 5
mc.getB(); // returns "passed"

var testClass = new TestClass();
mc.test(testClass);
mc.getA(); // returns 6
mc.getB(); // returns "something else"

mc.test("variadic", 1, 2, 5, "testing"); // returns "variadic = 1-2-5-testing"
```

### CODE REVIEW

File: dist/joii.js [hunk 1]
```diff
@@ -922,18 +1050,223 @@ JOII.CreateProperty = function(obj, name, val, writable) {
...
+    proto_meta.is_abstract = false;
+
+    for (var i = 0; i < meta.parameters.length - 1; i++) {
+        if (meta.parameters[i] == '...') {
+            throw 'Member ' + meta.name + ': Variadic parameter (...) must be the last in the function parameter list.';
+        }
+    }
+
+    for (var idx = 0; idx < proto_meta.overloads.length; idx++) {
+        var functionParametersMeta = proto_meta.overloads[idx];
```
Comment (haroldiedema) [inline dist/joii.js]: Please be consistent with codestyle in variable names.

- Variables are written in snake_case
- Classes (functions that can be instantiated using the `new` keyword)  are written in PascalCase (UpperCamelCase)
- Methods are written in camelCase.

The reason for this convention is so you can always see what a variable is without having to look it up.
This also applies to the rest of the commit.


I know a lot of the older code also doesn't fully comply to these rules (yet), but as the codebase grows, I think it's a good idea to comply to this sooner than later.

File: dist/joii.js [hunk 2]
```diff
@@ -922,18 +1050,223 @@ JOII.CreateProperty = function(obj, name, val, writable) {
...
+};
+
+
+/**
+ * Creates a function detour which checks type parameters, then dispatches to the appropriate overloaded function.
+ *
+ * @param {String}   name
+ * @param {Object}   overloads
+ */
+JOII.CreateFunctionShim = function(name, overloads) {
```
Comment (haroldiedema) [inline dist/joii.js]: Function isn't mean to be instantiated using the `new` keyword, therefore written in camelCase.

File: dist/joii.js [hunk 3]
```diff
@@ -1329,8 +1685,43 @@ JOII.InterfaceBuilder = function() {
...
-                    throw 'Method ' + p1.getName() + ' does not match the parameter count as defined in the interface ' + this.name + '.';
+                var argsInterface = p1.getParameters();
+                var argsClass = p2.getParameters();
+
+                if (argsInterface.length == 0 || typeof (argsInterface[0]) !== 'object') {
+                    // fallback for backwards compatibility
+                    if (p1.getParameters().length !== p2.getParameters().length) {
+                        throw 'Method ' + p1.getName() + ' does not match the parameter count as defined in the interface ' + this.name + '.';
+                    }
+                }
```
Comment (haroldiedema) [inline dist/joii.js]: codestyle.

File: test/ClassBuilder/CallTest.js [hunk 4]
```diff
@@ -40,7 +40,7 @@ test('ClassBuilder:CallTest', function(assert) {
     C2(2);
     var c2a = new C2();
     assert.strictEqual(C2(), 2, '__call returns this.a (2)');
-    assert.strictEqual(c2.getA(), 1, 'c2.getA() returns this.a (2)');
+    assert.strictEqual(c2.getA(), 1, 'c2.getA() returns this.a (2)'); // these are inconsistent. Either the output comment is wrong (2), or the test is wrong (1). Which was intended?
```
Comment (haroldiedema) [inline test/ClassBuilder/CallTest.js]: Nice catch!

So, there's actually 2 things wrong here. Actually one, but the other isn't implemented afaik.

1. The test is correct. The comment is wrong.
2. The "static" modifier isn't implemented. If I remember correctly, I couldn't find an efficient way to share a non-object variable over multiple instances of classes.

A static call function (__call), should NOT be able to overwrite the value of non-static properties in instantiated classes. However, If a property is declared as `static`, then it is allowed. Currently this isn't possible in JOII.

File: testsuite.js [hunk 5]
```diff
@@ -78,17 +95,47 @@ if (typeof(window) === 'undefined') {
...
+
+    var allScriptsToLoad = [];
+
+    var arrayIndex = 0;
+
+    if (typeof(testsuite.code) === 'object') {
+        totalScripts += testsuite.code.length;
+        allScriptsToLoad = testsuite.code.slice(0);
+    }
+    else {
```
Comment (haroldiedema) [inline testsuite.js]: codestyle

File: dist/joii.js (hunk 3 above)
Comment (cjmanca) [inline dist/joii.js]: Could you please explain the issue with this particular area for code style? I can't see anything wrong with it.

File: test/ClassBuilder/CallTest.js (hunk 4 above)
Comment (cjmanca) [inline test/ClassBuilder/CallTest.js]: I have a couple ideas on how to implement static members, but may need to break BC. I'll put some thought into how to do it without breaking BC.

**Comment by haroldiedema** [comment]
I absolutely love the addition! Thanks a lot for this!

Not sure what else you have in mind, but I would kindly suggest to keep the commits a tad smaller next time. It appears you applied some codestyle changes to all the code, making this a little difficult to review.

I'll wait for the fix in `InterfaceBuilder` @ line 90, then merge it 👍

**Comment by cjmanca** [comment]
VisualStudio kept trying to change a couple of spacing things. Eventually I got tired of doing a "replace all" to revert them each time, and just left them, lol.

**Comment by haroldiedema** [comment]
I removed the changes in testsuite.js, as the raw source doesn't work in neither Windows nor Linux. At first I thought it was a windows problem, but reverting the changes in testsuite.js back to the old one fixed the issue.

The `code` element in the `testsuite` object doesn't accept an array of files for me. I've tried this in the past, but to no avail.


I don't see how you got this to work. You should run `npm test` and refresh your browser page to ensure the compiled file works correctly, since that's what people are using.

**Comment by cjmanca** [comment]
Strange... working fine for me. I've tested it both ways. Perhaps a difference in the versions of qunit?

**Pull request merged**